File: packages/@emulators/okta/src/store.ts
import { Store, type Collection } from "@emulators/core";
import type {
  OktaUser,
  OktaGroup,
  OktaApp,
  OktaOAuthClient,
  OktaAuthorizationServer,
  OktaGroupMembership,
  OktaAppAssignment,
} from "./entities.js";

export interface OktaStore {
  users: Collection<OktaUser>;
  groups: Collection<OktaGroup>;
  apps: Collection<OktaApp>;
  oauthClients: Collection<OktaOAuthClient>;
  authorizationServers: Collection<OktaAuthorizationServer>;
  groupMemberships: Collection<OktaGroupMembership>;
  appAssignments: Collection<OktaAppAssignment>;
}

export function getOktaStore(store: Store): OktaStore {
  return {
    users: store.collection<OktaUser>("okta.users", ["okta_id", "login", "email"]),
    groups: store.collection<OktaGroup>("okta.groups", ["okta_id", "name"]),
    apps: store.collection<OktaApp>("okta.apps", ["okta_id", "name"]),
    oauthClients: store.collection<OktaOAuthClient>("okta.oauth_clients", ["client_id", "auth_server_id"]),
    authorizationServers: store.collection<OktaAuthorizationServer>("okta.auth_servers", ["server_id"]),
    groupMemberships: store.collection<OktaGroupMembership>("okta.group_memberships", [
      "group_okta_id",
      "user_okta_id",
    ]),
    appAssignments: store.collection<OktaAppAssignment>("okta.app_assignments", ["app_okta_id", "user_okta_id"]),
  };
}

File: internal/services/okta/store.go
package okta

import corestore "github.com/vercel-labs/emulate/internal/core/store"

type Store struct {
	Users                *corestore.Collection
	Groups               *corestore.Collection
	Apps                 *corestore.Collection
	OAuthClients         *corestore.Collection
	AuthorizationServers *corestore.Collection
	GroupMemberships     *corestore.Collection
	AppAssignments       *corestore.Collection
	OAuthCodes           *corestore.Collection
	AccessTokens         *corestore.Collection
	RefreshTokens        *corestore.Collection
}

func NewStore(store *corestore.Store) Store {
	return Store{
		Users:                store.MustCollection("okta.users", "okta_id", "login", "email"),
		Groups:               store.MustCollection("okta.groups", "okta_id", "name"),
		Apps:                 store.MustCollection("okta.apps", "okta_id", "name"),
		OAuthClients:         store.MustCollection("okta.oauth_clients", "client_id", "auth_server_id"),
		AuthorizationServers: store.MustCollection("okta.auth_servers", "server_id"),
		GroupMemberships:     store.MustCollection("okta.group_memberships", "group_okta_id", "user_okta_id"),
		AppAssignments:       store.MustCollection("okta.app_assignments", "app_okta_id", "user_okta_id"),
		OAuthCodes:           store.MustCollection("okta.oauth_codes", "code", "auth_server_id", "client_id", "user_okta_id"),
		AccessTokens:         store.MustCollection("okta.access_tokens", "token", "auth_server_id", "client_id", "user_okta_id"),
		RefreshTokens:        store.MustCollection("okta.refresh_tokens", "token", "auth_server_id", "client_id", "user_okta_id"),
	}
}
